core/src/main/scala/kafka/tools/MirrorMaker.scala

```diff
@@ -415,31 +323,41 @@ object MirrorMaker extends Logging with KafkaMetricsGroup {
 
     setName(threadName)
 
+    private def toBaseConsumerRecord(record: ConsumerRecord[Array[Byte], Array[Byte]]): BaseConsumerRecord =
+      BaseConsumerRecord(record.topic,
+        record.partition,
+        record.offset,
+        record.timestamp,
+        record.timestampType,
+        record.key,
+        record.value,
+        record.headers)
+
     override def run() {
       info("Starting mirror maker thread " + threadName)
       try {
-        mirrorMakerConsumer.init()
+        consumerWrapper.init()
 
-        // We need the two while loop to make sure when old consumer is used, even there is no message we
-        // still commit offset. When new consumer is used, this is handled by poll(timeout).
+        // FIXME Is it safe to remove one of the loops?
         while (!exitingOnSendFailure && !shuttingDown) {
           try {
-            while (!exitingOnSendFailure && !shuttingDown && mirrorMakerConsumer.hasData) {
-              val data = mirrorMakerConsumer.receive()
+            while (!exitingOnSendFailure && !shuttingDown) {
+              val data = consumerWrapper.receive()
               if (data.value != null) {
                 trace("Sending message with value size %d and offset %d.".format(data.value.length, data.offset))
               } else {
                 trace("Sending message with null value and offset %d.".format(data.offset))
               }
-              val records = messageHandler.handle(data)
+              val records = messageHandler.handle(toBaseConsumerRecord(data))
               records.asScala.foreach(producer.send)
               maybeFlushAndCommitOffsets()
             }
           } catch {
-            case _: ConsumerTimeoutException =>
-              trace("Caught ConsumerTimeoutException, continue iteration.")
+            //FIXME Is this change OK?
```

Comment:
I think this change is OK.

Reply:
Thanks. What I was wondering about is whether there were cases where `TimeoutException` is thrown that we don't want to catch and continue. `ConsumerTimeoutException` is only thrown by `NewShinyConsumer` if `poll` doesn't return any records. Maybe to be safe, I should maintain the exact behaviour and then we can a separate PR to simplify MirrorMaker (instead of being in the middle of this huge one).